### PR TITLE
implement async lock to replace redis locks for testing

### PR DESCRIPTION
**Pull request by pvanderlinden**

### CODE REVIEW

File: lux/core/cache.py [hunk 1]
```diff
@@ -73,14 +74,101 @@ def lock(self, name, timeout=None):
         raise NotImplementedError
 
 
+class AsyncLock:
+
+    _locks = {}
+
+    def __init__(self, name, loop=None, timeout=None, blocking=0, sleep=0.2):
```
Comment (lsbardel) [inline lux/core/cache.py]:
```python
self._loop = loop
```

File: lux/core/cache.py [hunk 2]
```diff
@@ -73,14 +74,101 @@ def lock(self, name, timeout=None):
...
+            self._lock.release()
+            self._cancel_timeout()
+        except RuntimeError as exc:
+            raise LockError(str(exc)) from exc
+        finally:
+            self._acquired = False
+
+    @classmethod
+    def _get_lock(cls, name, loop=None):
+        if not name in cls._locks:
```
Comment (lsbardel) [inline lux/core/cache.py]: what about if we have a lock name for different loops?

File: lux/core/cache.py [hunk 3]
```diff
@@ -73,14 +74,101 @@ def lock(self, name, timeout=None):
...
+            cls._locks[name] = asyncio.Lock(loop=loop)
+        return cls._locks[name]
+
+    def _schedule_timeout(self):
```
Comment (lsbardel) [inline lux/core/cache.py]: I would not use async sleep here. I would use the event loop directly:
```python
self._timeout_handler = self._loop.call_later(self._timeout, self._release_after_timeout)
```

File: lux/core/cache.py [hunk 4]
```diff
@@ -73,14 +74,101 @@ def lock(self, name, timeout=None):
...
+            return
+        if not self._timeout_future.done():
+            self._timeout_future.cancel()
+        self._timeout_future = None
+
+    @asyncio.coroutine
+    def _release_after_timeout(self):
+        yield from asyncio.sleep(self._timeout)
+        if self._acquired:
+            self.release()
```
Comment (lsbardel) [inline lux/core/cache.py]:
```python
def _release_after_timeout(self):
    if self._acquired:
        self.release()
```

File: lux/core/cache.py [hunk 5]
```diff
@@ -73,14 +74,101 @@ def lock(self, name, timeout=None):
...
+        return cls._locks[name]
+
+    def _schedule_timeout(self):
+        if self._timeout is None:
+            return
+        self._timeout_future = asyncio.async(self._release_after_timeout())
+
+    def _cancel_timeout(self):
+        if self._timeout is None or self._timeout_future is None:
+            return
```
Comment (lsbardel) [inline lux/core/cache.py]:
```python
self._timeout_handler.cancel()
self._timeout_handler = None
```

File: lux/core/cache.py [hunk 6]
```diff
@@ -73,14 +74,101 @@ def lock(self, name, timeout=None):
...
 
+class AsyncLock:
+
+    _locks = {}
+
+    def __init__(self, name, loop=None, timeout=None, blocking=0, sleep=0.2):
+        self._lock = self._get_lock(name, loop)
+        self._timeout = timeout
+        self._blocking = blocking
+        self._acquired = False
```
Comment (lsbardel) [inline lux/core/cache.py]: self._timeout_handler = None

File: lux/core/cache.py (hunk 1 above)
Comment (pvanderlinden) [inline lux/core/cache.py]: Fixed

**Comment by pvanderlinden** [comment]
Not to sure what it needs to do with multiple loops but the same name for the lock. One possibility would be asserting that the loop for the to is the same, and raise a LockError if that fails, or let them live next to each other and see them as separate locks. Probably the first solution would be best (no unexpected failures)?

**Comment by lsbardel** [comment]
what about having the ``_locks`` dictionary injected in the event loop?
You can use several loops for testing reasons.

**Comment by pvanderlinden** [comment]
Implemented the _locks per loop now

**Comment by lsbardel** [comment]
looks good

**Pull request merged**